Hw10 program optimization

director74 opened the pull request:
## Домашнее задание №10 «Оптимизация программы»

### Чек-лист студента ([Что это?](https://github.com/OtusGolang/home_work/wiki/%D0%9A%D0%BE%D0%BC%D0%BC%D0%B5%D0%BD%D1%82%D0%B0%D1%80%D0%B8%D0%B8-%D0%BA-%D1%87%D0%B5%D0%BA-%D0%BB%D0%B8%D1%81%D1%82%D1%83-%D1%81%D1%82%D1%83%D0%B4%D0%B5%D0%BD%D1%82%D0%B0))
- [x] Я перечитал задание после решения. [Зачем перечитывать задание?](https://github.com/OtusGolang/home_work/wiki/%D0%9A%D0%BE%D0%BC%D0%BC%D0%B5%D0%BD%D1%82%D0%B0%D1%80%D0%B8%D0%B8-%D0%BA-%D1%87%D0%B5%D0%BA-%D0%BB%D0%B8%D1%81%D1%82%D1%83-%D1%81%D1%82%D1%83%D0%B4%D0%B5%D0%BD%D1%82%D0%B0#user-content-%D0%97%D0%B0%D1%87%D0%B5%D0%BC-%D0%BF%D0%B5%D1%80%D0%B5%D1%87%D0%B8%D1%82%D1%8B%D0%B2%D0%B0%D1%82%D1%8C-%D0%B7%D0%B0%D0%B4%D0%B0%D0%BD%D0%B8%D0%B5)
- [x] Я запустил `go mod tidy`.
- [x] Я удалил `.sync` файл. [Зачем его удалять?](https://github.com/OtusGolang/home_work/wiki/%5B%D0%A1%D1%82%D1%83%D0%B4%D0%B5%D0%BD%D1%82%D0%B0%D0%BC%5D-%D0%9F%D1%80%D0%BE%D1%86%D0%B5%D1%81%D1%81-%D1%81%D0%B4%D0%B0%D1%87%D0%B8-%D0%94%D0%97#user-content-%D0%92%D0%B0%D1%80%D0%B8%D0%B0%D0%BD%D1%82-2)
- [x] Я написал бенчмарк и приложил результаты benchstat.

### Критерии оценки
- [x] Пайплайн зелёный и нет попытки «обмануть» систему - 4 балла
- [x] Добавлены юнит-тесты - до 3 баллов
- [x] Понятность и чистота кода - до 3 баллов

#### Зачёт от 7 баллов

Commit:
HW9 validator realisation

Commit:
HW9 linter fixes

Commit:
HW9 test cases

Commit:
HW9 test cases

Commit:
Merge branch 'master' into hw09_struct_validator

Commit:
HW9 fixes

Commit:
HW9 fixes

Commit:
HW10 in process

Commit:
HW10 acceleration

Commit:
HW10 benchmark result

name             old time/op    new time/op    delta
GetDomainStat-3    7.66ms ±19%    6.93ms ± 1%   -9.50%  (p=0.023 n=10+10)

name             old alloc/op   new alloc/op   delta
GetDomainStat-3    2.15MB ±11%    0.03MB ± 0%  -98.73%  (p=0.000 n=8+9)

name             old allocs/op  new allocs/op  delta
GetDomainStat-3     22.1k ±11%      0.9k ± 0%  -96.02%  (p=0.000 n=8+9)

Commit:
HW10 additional tests

Commit:
HW10 linter

director74 commented:
результаты бенчмарка

name             old time/op    new time/op    delta
GetDomainStat-3    7.66ms ±19%    6.93ms ± 1%   -9.50%  (p=0.023 n=10+10)

name             old alloc/op   new alloc/op   delta
GetDomainStat-3    2.15MB ±11%    0.03MB ± 0%  -98.73%  (p=0.000 n=8+9)

name             old allocs/op  new allocs/op  delta
GetDomainStat-3     22.1k ±11%      0.9k ± 0%  -96.02%  (p=0.000 n=8+9)

agneum commented on hw09_struct_validator/go.mod:
```diff
@@ -1,3 +1,5 @@
-module github.com/fixme_my_friend/hw09_struct_validator
+module github.com/director74/homeworks/hw09_struct_validator
```
Просьба, чтобы PR содержал только файлы, относящиеся к текущему ДЗ

director74 replied:
прошу прощения, так по ошибке получилось

agneum commented on hw10_program_optimization/stats.go:
```diff
@@ -32,32 +31,26 @@ func GetDomainStat(r io.Reader, domain string) (DomainStat, error) {
+			return result, err
 		}
-		result[i] = user
+		result[i] = *user
```
В рамках задачи можно не сохранять данные пользователей и сразу подсчитывать количество доменов

director74 replied:
не совсем понял
у меня тут user только с одним полем Email сохраняется, другие поля убирал из UnmarshalJSON
про эту оптимизацию речь?

director74 replied:
теперь понял, убрал лишнюю функцию и сохранение - стало еще быстрее!
спасибо

agneum commented on hw10_program_optimization/stats.go:
```diff
@@ -32,32 +31,26 @@ func GetDomainStat(r io.Reader, domain string) (DomainStat, error) {
-		}
-
-		if matched {
+		if strings.Contains(user.Email, "."+domain) {
```
Предложил бы использовать `strings.HasSuffix`

director74 replied:
поменял на strings.HasSuffix

agneum commented on hw10_program_optimization/stats_test.go:
```diff
@@ -36,4 +39,30 @@ func TestGetDomainStat(t *testing.T) {
+
+func BenchmarkGetDomainStat(b *testing.B) {
+	r, _ := zip.OpenReader("testdata/users.dat.zip")
+	data, _ := r.File[0].Open()
```
В тестах ошибки важно тоже обрабатывать

director74 replied:
добавил,
думал что в бенчмарке каким-то хитрым способом проверяется, потому что нет testing.T

agneum commented on hw10_program_optimization/stats_test.go:
```diff
@@ -36,4 +39,30 @@ func TestGetDomainStat(t *testing.T) {
+
+func BenchmarkGetDomainStat(b *testing.B) {
+	r, _ := zip.OpenReader("testdata/users.dat.zip")
+	data, _ := r.File[0].Open()
```
Рекомендую использовать https://pkg.go.dev/testing#B.StopTimer и https://pkg.go.dev/testing#B.StartTimer, чтобы исключить время на подготовку данных

agneum reviewed (approved):
Спасибо за выполненное задание. Хорошая работа
Итого: 10 баллов из 10.

Commit:
HW10 upgrades

Commit:
HW10 upgrades

Commit:
HW10 new benchmark comparison
name             old time/op    new time/op    delta
GetDomainStat-3    7.66ms ±19%    0.00ms ± 7%  -99.99%  (p=0.000 n=10+9)

name             old alloc/op   new alloc/op   delta
GetDomainStat-3    2.15MB ±11%    0.00MB ± 0%  -99.81%  (p=0.000 n=8+10)

name             old allocs/op  new allocs/op  delta
GetDomainStat-3     22.1k ±11%      0.0k ± 0%  -99.99%  (p=0.000 n=8+10)
